libsplinter/src/service/scabbard/client/mod.rs

```diff
@@ -0,0 +1,311 @@
+// Copyright 2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod contract_registry;
+mod error;
+mod execute;
+mod key;
+mod namespace;
+mod smart_permission;
+mod submit;
+mod transaction;
+mod upload;
+
+use std::fs::File;
+use std::io::{BufReader, Read};
+
+use sabre_sdk::protocol::payload::SabrePayload;
+pub use sawtooth_sdk::signing::secp256k1::Secp256k1PrivateKey;
+
+pub use error::Error;
+use submit::submit_batch_list;
+use transaction::batch_payloads;
+pub use upload::{SmartContract, SmartContractMetadata};
+
+/// A client that can be used to submit transactions to a scabbard service.
+pub struct ScabbardClient {
+    url: String,
+    signing_key: Secp256k1PrivateKey,
+    unsent_payloads: Vec<SabrePayload>,
+}
+
+impl ScabbardClient {
+    /// Create a new `ScabbardClient` with the given `url` and a signing key loaded from the given
+    /// `key_file`.
+    pub fn new_with_local_signing_key(url: &str, key_file: &str) -> Result<Self, Error> {
+        Ok(Self::new(url, Self::load_signing_key_from_file(key_file)?))
+    }
+
+    /// Create a new `ScabbardClient` with the given `url` and `signing_key`.
+    pub fn new(url: &str, signing_key: Secp256k1PrivateKey) -> Self {
+        Self {
+            url: url.into(),
+            signing_key,
+            unsent_payloads: vec![],
+        }
+    }
+
+    /// Submit all unsent transactions to the scabbard service.
+    pub fn submit(mut self) -> Result<(), Error> {
+        let batch_list = batch_payloads(&mut self.unsent_payloads.drain(..), &self.signing_key)?;
+        let _batch_link = submit_batch_list(&self.url, &batch_list)?;
+        Ok(())
+    }
+
+    /// Queue a transaction to upload a smart contract loaded from the given `.scar`. See
+    /// `load_scar` for more on how the `.scar` is loaded.
+    pub fn upload_contract_from_scar(self, scar: &str) -> Result<Self, Error> {
```

Comment:
It may be useful to take a more nuanced approach to this where not only is the .scar loaded and the upload transaction added, but it also adds the related transactions for the contract (e.g. creating registries and permissions).